### PR TITLE
Make allocate_set(s) failable

### LINKED ISSUE #1922: Make `allocate_set`/`allocate_sets` failable
The descriptor pool is limited, we can run out of descriptors.

This is kind of a blocker for VkPortability, since we are getting an out of descriptor error that's not handled properly by `ash`, and we need to forward it up instead of unwrapping. cc @MaikKlein

### PR DESCRIPTION
**Pull request by AIOOB**
Fixes #1922 
PR checklist:
- [x] `make` succeeds (on *nix)
- [ ] `make reftests` succeeds
- [ ] tested examples with the following backends:

### CODE REVIEW

File: src/hal/src/pso/descriptor.rs [hunk 1]
```diff
@@ -87,26 +87,42 @@ pub struct DescriptorRangeDesc {
...
+#[derive(Fail, Clone, Copy, Debug, PartialEq)]
+pub enum AllocationError {
+    /// Memory allocation on the host side failed.
+    /// This could be caused by a lack of memory or pool fragmentation.
+    #[fail(display = "Host memory allocation failed.")]
+    OutOfHostMemory,
+    /// Memory allocation on the host side failed.
+    /// This could be caused by a lack of memory or pool fragmentation.
+    #[fail(display = "Device memory allocation failed.")]
+    OutOfDeviceMemory,
```
Comment (AIOOB) [inline src/hal/src/pso/descriptor.rs]: The vulkan spec is unclear to me as to whether these errors mean something different to the fragmented pool error as I would assume it's possible to run out of memory without having a fragmented pool.

File: src/hal/src/pso/descriptor.rs [hunk 2]
```diff
@@ -87,26 +87,42 @@ pub struct DescriptorRangeDesc {
...
     }
 
     /// Allocate one or multiple descriptor sets from the pool.
     ///
     /// Each descriptor set will be allocated from the pool according to the corresponding set layout.
-    /// The descriptor pool _must_ have enough space in to allocate the required descriptors.
     /// Descriptors will become invalid once the pool is reset. Usage of invalidated descriptor sets results
     /// in undefined behavior.
-    fn allocate_sets<I>(&mut self, layouts: I) -> Vec<B::DescriptorSet>
+    fn allocate_sets<I>(&mut self, layouts: I) -> Vec<Result<B::DescriptorSet, AllocationError>>
```
Comment (AIOOB) [inline src/hal/src/pso/descriptor.rs]: Vulkan provides a Result<Vec> but I think a Vec<Result> makes a bit more sense unless we want to promise that the length of the output and the length of the input will be equal.

File: src/backend/metal/src/native.rs [hunk 3]
```diff
@@ -165,24 +165,24 @@ impl hal::DescriptorPool<Backend> for DescriptorPool {
...
             DescriptorPool::ArgumentBuffer { ref buffer, total_size, ref mut offset } => {
                 let (encoder, stage_flags) = match layout {
                     &DescriptorSetLayout::ArgumentBuffer(ref encoder, stages) => (encoder, stages),
-                    _ => panic!("Incompatible descriptor set layout type"),
+                    _ => return Err(pso::AllocationError::IncompatibleLayout),
                 };
 
                 let cur_offset = *offset;
                 *offset += encoder.encoded_length();
                 assert!(*offset <= total_size);
```
Comment (kvark) [inline src/backend/metal/src/native.rs]: this is where we should return OOM error instead of asserting

File: src/backend/dx12/src/native.rs [hunk 4]
```diff
@@ -447,13 +447,13 @@ impl HalDescriptorPool<Backend> for DescriptorPool {
             };
         }
 
-        DescriptorSet {
+        Ok(DescriptorSet {
```
Comment (kvark) [inline src/backend/dx12/src/native.rs]: the OOM error would go from failing `alloc_handles`, although it's fine to do later on

**Comment by AIOOB** [comment]
bors r=kvark

**Pull request merged**